File: Algorithms/searching/ts/depth_first_search.ts
class Graph3 {
  adjacencyList: Map<number, number[]> = new Map();

  constructor() {}

  // Add a vertex to the graph
  addVertex(vertex: number): void {
    this.adjacencyList.set(vertex, []);
  }

  // Add an edge between two vertices
  addEdge(source: number, destination: number): void {
    const neighbors = this.adjacencyList.get(source);
    if (neighbors) {
      neighbors.push(destination);
    }
  }

  // Perform a depth-first search starting from the given vertex
  depthFirstSearch(startVertex: number): void {
    const visited: Set<number> = new Set();
    this.depthFirstSearchRecursive(startVertex, visited);
  }

  // Recursive helper function for depthFirstSearch
  private depthFirstSearchRecursive(
    vertex: number,
    visited: Set<number>
  ): void {
    visited.add(vertex);
    console.log(vertex);

    const neighbors = this.adjacencyList.get(vertex) || [];
    for (const neighbor of neighbors) {
      if (!visited.has(neighbor)) {
        this.depthFirstSearchRecursive(neighbor, visited);
      }
    }
  }
}

// Usage Example
const graph3 = new Graph3();

graph3.addVertex(0);
graph3.addVertex(1);
graph3.addVertex(2);
graph3.addVertex(3);
graph3.addVertex(4);

graph3.addEdge(0, 1);
graph3.addEdge(0, 2);
graph3.addEdge(1, 3);
graph3.addEdge(2, 4);

console.log("Depth First Search starting from vertex 0:");
graph3.depthFirstSearch(0);

File: Algorithms/searching/dart/depth_first_search.dart
void main() {
  Graph graph = Graph();

  graph.addVertex(0);
  graph.addVertex(1);
  graph.addVertex(2);
  graph.addVertex(3);
  graph.addVertex(4);

  graph.addEdge(0, 1);
  graph.addEdge(0, 2);
  graph.addEdge(1, 3);
  graph.addEdge(2, 4);

  print("Depth First Search starting from vertex 0:");
  graph.depthFirstSearch(0);
}

class Graph {
  Map<int, List<int>> adjacencyList = {};

  Graph();

  // Add a vertex to the graph
  void addVertex(int vertex) {
    adjacencyList[vertex] = [];
  }

  // Add an edge between two vertices
  void addEdge(int source, int destination) {
    adjacencyList[source]?.add(destination);
  }

  // Perform a depth-first search starting from the given vertex
  void depthFirstSearch(int startVertex) {
    Set<int> visited = {};
    depthFirstSearchRecursive(startVertex, visited);
  }

  // Recursive helper function for depthFirstSearch
  void depthFirstSearchRecursive(int vertex, Set<int> visited) {
    visited.add(vertex);
    print(vertex);

    for (int neighbor in adjacencyList[vertex] ?? []) {
      if (!visited.contains(neighbor)) {
        depthFirstSearchRecursive(neighbor, visited);
      }
    }
  }
}
